src/components/NetworkModal/index.tsx

```diff
@@ -229,22 +236,32 @@ export default function NetworkModal(): JSX.Element | null {
                                     ].includes(key as any)
                                 ) {
                                     console.log(key.toString(16))
-                                    if (isMetaMask) {
+                                    if (isMetaMask && isOnlyMeta) {
                                         ; (ethereum as any).request({
                                             method: 'wallet_switchEthereumChain',
                                             params: [{ chainId: `0x${key.toString(16)}` }]
                                         })
+                                    } else if (isMetaMask && !isOnlyMeta) {
+                                      ; (ethereum?.selectedProvider as any).request({
+                                        method: 'wallet_switchEthereumChain',
+                                        params: [{ chainId: `0x${key.toString(16)}` }]
+                                      })
                                     } else {
                                         library?.send('wallet_switchEthereumChain', [
                                             { chainId: `0x${key.toString(16)}` }
                                         ])
                                     }
                                 } else {
-                                    if (isMetaMask) {
+                                    if (isMetaMask && isOnlyMeta) {
                                         ; (ethereum as any).request({
                                             method: 'wallet_addEthereumChain',
                                             params: [params, account]
                                         })
+                                      } else if (isMetaMask && !isOnlyMeta) {
+                                        ; (ethereum?.selectedProvider as any).request({
+                                          method: 'wallet_addEthereumChain',
+                                          params: [params, account]
+                                        })
```

Comment:
we always put the provider in ethereum.selectedProvider so no need for duplicate code. also all conditions are not required. no need for ethereum? as any condition (isMetamask or ethereum) already assumes ethereum exists.
The only condition that is required is `isMetaMask` because wallet_X API is specific to metamask wallet.

Reply:
I was looking over it, and yes I agree it's here and there repeated
but I wasn't gonna fine-tune anything until I knew what our preferred direction would be.

That said.

First comment: no it only sets the selectedProvider (for now) if multiple providers exist, else it goes as before
(which is what we want, as then all the provider logic is as the window interface defines it)
if multiple providers do exists, the scopes change
so we could decide to always set the selectedProvider, but then I have to change a little more of the original flow

Comment 2: I could change it to put ethereum inside the IF statement
but the code cannot compile if I didn't put ethereum? in (object possibly undefined..)

And about the last part

CoinBase injects itself into the window.ethereum, in a way that the Top-Level isMetaMask is ALWAYS true
now we have to dive inside the different providers to know which one it is

so the top-level one is no indication for anything (if multiple providers exist)

Reply:
what i said was based on firstly applying my code suggestion